sklearn/utils/_sorting.pxd

```diff
@@ -0,0 +1,19 @@
+from ._typedefs cimport DTYPE_t, ITYPE_t
+
+from cython cimport floating
+
+cdef int partition_node_indices(
+    DTYPE_t *data,
+    ITYPE_t *node_indices,
+    ITYPE_t split_dim,
+    ITYPE_t split_index,
+    ITYPE_t n_features,
+    ITYPE_t n_points
+) nogil except -1
```

Comment:
I do not think this will be used in others implementation than `BinaryTree`.

I moved `partition_node_indices` here to have it be part of a private module defining interfaces to `std::algorithm` and relevant `Comparators`.

Note that this is still a draft: I started exploring data-structures and sorting algorithms in the code-base to see what could fit here.

Reply:
We have an different inclusion criteria for the utils module. My criteria is: "Is the utility used in multiple modules?" I do not mind having `std::algorithm` interfaces in different modules.

I like `partition_node_indices` in `neighbors` because it is scoped to the `neighbors` module.

Reply:
OK let's revert it.